builder/pipeline_test.cc

```diff
@@ -1,15 +1,71 @@
 #include <gtest/gtest.h>
 
 #include <cassert>
+#include <fstream>
 
 #include "builder/pipeline.h"
+#include "builder/replica_pipeline.h"
 #include "runtime/expr.h"
 #include "runtime/pipeline.h"
 #include "runtime/thread_pool.h"
 #include "runtime/visualize.h"
+#include "tools/cpp/runfiles/runfiles.h"
 
 namespace slinky {
 
+namespace internal {
+extern void check_replica_pipeline(const std::string& name, const std::string& replica_text,
+    span<const raw_buffer*> inputs, span<const raw_buffer*> outputs, const std::string& test_src);
+}  // namespace internal
+
+using bazel::tools::cpp::runfiles::Runfiles;
+
+std::string read_entire_file(const std::string& pathname) {
+  std::ifstream f(pathname, std::ios::in | std::ios::binary);
+  std::string result;
+
+  f.seekg(0, std::ifstream::end);
+  size_t size = f.tellg();
+  result.resize(size);
+  f.seekg(0, std::ifstream::beg);
+  f.read(result.data(), result.size());
+  if (!f.good()) {
+    std::cerr << "Unable to read file: " << pathname;
+    std::abort();
+  }
+  f.close();
+  return result;
+}
+
+class PipelineTest : public testing::Test {
+protected:
+  void SetUp() override {
+    // TODO: for testing purposes, remove
+    // replica_pipeline_test_src = read_entire_file("/Users/srj/GitHub/slinky/builder/replica_pipeline_test.cc");
+    // return;
+
+    std::string error;
+    runfiles.reset(Runfiles::CreateForTest(BAZEL_CURRENT_REPOSITORY, &error));
+    if (runfiles == nullptr) {
+      std::cerr << "Can't find runfile directory: " << error;
+      std::abort();
+    }
+
+    // As of Bazel 6.x, apparently `_main` is the toplevel for runfiles
+    // (https://github.com/bazelbuild/bazel/issues/18128)
+    auto full_path = runfiles->Rlocation("_main/builder/replica_pipeline_test.cc");
+    replica_pipeline_test_src = read_entire_file(full_path);
+  }
+
+  void check_replica_pipeline(const std::string& name, const std::string& replica_text, span<const raw_buffer*> inputs,
+      span<const raw_buffer*> outputs) {
+    ::slinky::internal::check_replica_pipeline(name, replica_text, inputs, outputs, replica_pipeline_test_src);
```

Comment:
This is still going further than I expected, which is that in `pipeline_test.cc`, literally the only thing we do is check that the replica pipeline is a substring. The actual test would be in `replica_pipeline_test.cc`.

Reply:
We have not communicated expectations well for this PR. I'll keep trying to comply with what you have in mind.